Include/XTaskQueue.h

```diff
@@ -0,0 +1,309 @@
+/// preventning new items from being queued.  Once a queue is terminated
+/// its handle can be closed. New items cannot be enqueued to a task
+/// queue that has been terminated.
+/// </summary>
+/// <param name='queue'>The queue to terminate.</param>
+/// <param name='wait'>True to wait for the termination to complete.</param>
+/// <param name='callbackContext'>An optional context pointer to pass to the callback.</param>
+/// <param name='callback'>An optional callback that will be called when the queue has terminated.</param>
+STDAPI XTaskQueueTerminate(
+    _In_ XTaskQueueHandle queue, 
+    _In_ bool wait, 
+    _In_opt_ void* callbackContext, 
+    _In_opt_ XTaskQueueTerminatedCallback* callback
+    ) noexcept;
+
+/// <summary>
+/// Submits a callback to the queue for the given port.
+/// </summary>
+/// <param name='queue'>The queue to submit the callback to.</param>
+/// <param name='port'>The port to submit the callback to.</param>
+/// <param name='callbackContext'>An optional context pointer that will be passed to the callback.</param>
+/// <param name='callback'>A pointer to the callback function.</param>
+STDAPI XTaskQueueSubmitCallback(
+    _In_ XTaskQueueHandle queue,
+    _In_ XTaskQueuePort port,
+    _In_opt_ void* callbackContext,
+    _In_ XTaskQueueCallback* callback
+    ) noexcept;
+
+/// <summary>
+/// Submits a callback to the queue for the given port.  The callback will be added
+/// to the queue after delayMs milliseconds.
+/// </summary>
+/// <param name='queue'>The queue to submit the callback to.</param>
+/// <param name='port'>The port to submit the callback to.</param>
+/// <param name='delayMs'>The number of milliseconds to delay before submitting the callback to the queue.</param>
+/// <param name='callbackContext'>An optional context pointer that will be passed to the callback.</param>
+/// <param name='callback'>A pointer to the callback function.</param>
+STDAPI XTaskQueueSubmitDelayedCallback(
+    _In_ XTaskQueueHandle queue,
+    _In_ XTaskQueuePort port,
+    _In_ uint32_t delayMs,
+    _In_opt_ void* callbackContext,
+    _In_ XTaskQueueCallback* callback
+    ) noexcept;
+
+/// <summary>
+/// Registers a wait handle with the task queue.  When the wait handle
+/// is satisfied the task queue will invoke the given callback. This
+/// provides an efficient way to add items to a task queue in 
+/// response to handles becomming signaled.
+/// </summary>
+/// <param name='queue'>The queue to submit the callback to.</param>
+/// <param name='port'>The port to invoke the callback on.</param>
+/// <param name='waitHandle'>The handle to monitor.</param>
+/// <param name='callbackContext'>An optional context pointer that will be passed to the callback.</param>
+/// <param name='callback'>A pointer to the callback function.</param>
+/// <param name='token'>A registration token.</param>
+STDAPI XTaskQueueRegisterWaiter(
+    _In_ XTaskQueueHandle queue,
+    _In_ XTaskQueuePort port,
+    _In_ HANDLE waitHandle,
```

Comment:
How does this work on non windows platforms?

Reply:
It returns E_NOTIMPL.